my-account/app/com/commercetools/sunrise/myaccount/authentication/recoverpassword/recover/DefaultRecoverPasswordControllerAction.java

```diff
@@ -0,0 +1,65 @@
+import io.sphere.sdk.customers.commands.CustomerCreatePasswordTokenCommand;
+import play.libs.concurrent.HttpExecution;
+
+import javax.inject.Inject;
+import javax.mail.Message;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.concurrent.CompletionStage;
+
+public class DefaultRecoverPasswordControllerAction extends AbstractCustomerCreatePasswordTokenExecutor implements RecoverPasswordControllerAction {
+    private RecoverPasswordEmailPageContentFactory recoverPasswordEmailPageContentFactory;
+    private final TemplateEngine templateEngine;
+    private final EmailSender emailSender;
+    private final Locale locale;
+
+    @Inject
+    protected DefaultRecoverPasswordControllerAction(final SphereClient sphereClient, final HookRunner hookRunner,
+                                                     final RecoverPasswordEmailPageContentFactory recoverPasswordEmailPageContentFactory,
+                                                     final TemplateEngine templateEngine,
+                                                     final EmailSender emailSender,
+                                                     final Locale locale) {
+        super(sphereClient, hookRunner);
+        this.recoverPasswordEmailPageContentFactory = recoverPasswordEmailPageContentFactory;
+        this.templateEngine = templateEngine;
+        this.emailSender = emailSender;
+        this.locale = locale;
+    }
+
+    @Override
+    public CompletionStage<CustomerToken> apply(final RecoverPasswordFormData recoveryEmailFormData) {
+        return executeRequest(buildRequest(recoveryEmailFormData))
+                .thenComposeAsync(customerToken -> onResetPasswordTokenCreated(customerToken, recoveryEmailFormData), HttpExecution.defaultContext());
+    }
+
+    protected CustomerCreatePasswordTokenCommand buildRequest(final RecoverPasswordFormData formData) {
+        return CustomerCreatePasswordTokenCommand.of(formData.email());
+    }
+
+    protected CompletionStage<CustomerToken> onResetPasswordTokenCreated(final CustomerToken resetPasswordToken, final RecoverPasswordFormData recoveryEmailFormData) {
+        final RecoverPasswordEmailPageContent recoverPasswordEmailPageContent =
+                recoverPasswordEmailPageContentFactory.create(resetPasswordToken);
+        final PageData pageData = new PageData();
+        pageData.setContent(recoverPasswordEmailPageContent);
+
+        final TemplateContext templateContext = new TemplateContext(pageData, Collections.singletonList(locale), null);
+        final String emailText = templateEngine.render("password-reset-email", templateContext);
+
+        return emailSender.send(msg -> {
+            msg.setRecipients(Message.RecipientType.TO, recoveryEmailFormData.email());
+            msg.setSubject(recoverPasswordEmailPageContent.getSubject(), "UTF-8");
+            msg.setContent(emailText, "text/html");
+        }).thenApply(s -> resetPasswordToken);
```

Comment:
I just thought this block here is a perfect candidate to go into its own class: building the page data, rendering the html content, sending the email. Then the tests would definitely be cleaner :)
That's the part we were missing I think. What do you think?
In any case, it's a test, we can improve it later.